


<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Visualization of microbial genomes &#8212; Microbial Genomics Lab Spring 2022  documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/cloud.css" />
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Noticia+Text:400,i,b,bi|Open+Sans:400,i,b,bi|Roboto+Mono:400,i,b,bi&amp;display=swap" type="text/css" />
    
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>

    
    
     
        <script src="_static/jquery.cookie.js"></script>
    

    
     
        <script src="_static/cloud.base.js"></script>
    

    
     
        <script src="_static/cloud.js"></script>
    

    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Week 12" href="Week12.html" /> 
        <meta name="viewport" content="width=device-width, initial-scale=1">
  </head><body>
    <div class="relbar-top">
        
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="Week12.html" title="Week 12"
             accesskey="P">previous</a> &nbsp; &nbsp;</li>
    <li><a href="index.html">Microbial Genomics Lab Spring 2022  documentation</a> &#187;</li>

        <li class="nav-item nav-item-this"><a href="">Visualization of microbial genomes</a></li> 
      </ul>
    </div>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for cloud theme */

/* nicer titles and more space for info and warning logos */

div.admonition p.admonition-title {
    background: rgba(0, 0, 0, .05);
    margin: .5em -1em;
    margin-top: -.5em !important;
    padding: .5em .5em .5em 2.65em;
}

/* indent single paragraph */
div.admonition {
    text-indent: 20px;
}
/* don't indent multiple paragraphs */
div.admonition > p {
    text-indent: 0;
}
/* remove excessive padding */
div.admonition.inline-title p.admonition-title {
    padding-left: .2em;
}
</style>
<div class="section" id="Visualization-of-microbial-genomes">
<h1>Visualization of microbial genomes<a class="headerlink" href="#Visualization-of-microbial-genomes" title="Permalink to this headline">¶</a></h1>
<p>There are various tools and techniques for visualizing microbial genomes. Take for example, this figure from a paper I published in <a class="reference external" href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0076376">2013</a>.</p>
<img alt="Genome Circle" src="_images/plos_one_circle.png" />
<p>This figure was created using a tool known as <strong>Circos</strong>. See here: <a class="reference external" href="http://circos.ca/">http://circos.ca/</a></p>
<p>If you visit this page, you can see beautiful pictures of various circular drawings produced using this tool. See here for examples of published images created with this tool: <a class="reference external" href="http://circos.ca/intro/published_images/">http://circos.ca/intro/published_images/</a></p>
<p>They have highlighted the figure I created on this page as well. This tool has been used not only in genomics but other fields such as economics and political science, for example. This tool has a pretty steep learning curve and it will take a bit of time to learn to use this tool. However, it comes with various tutorials and example data that you can start with. Then you can learn to adapt it to visualize your own data.</p>
<p>You can visit this page (<a class="reference external" href="http://circos.ca/documentation/">http://circos.ca/documentation/</a>) to download Circos tutorials to try their examples. Circos is available through Bioconda so you should be able to install it using the <code class="docutils literal notranslate"><span class="pre">conda</span></code> command.</p>
<p>Circos has been adapted into R packages as well. See <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-244">here</a> and <a class="reference external" href="https://cran.r-project.org/web/packages/BioCircos/vignettes/BioCircos.html">here</a>. A few known tools similar to or related to the original Circos are, <a class="reference external" href="https://cran.r-project.org/web/packages/RCircos/">R Circos</a>, <a class="reference external" href="https://cran.r-project.org/web/packages/BioCircos/vignettes/BioCircos.html">BioCircos</a>, and
<a class="reference external" href="https://jokergoo.github.io/circlize_book/book/">Circlize</a>. R versions of these tools make it a lot easier for you to create circular visualizations of microbial genomes, especially those packages with clear examples to follow. I highly recommend you check out the circlize R package as it provides very detailed instructions and examples to follow.</p>
<p>We will not be learning to use all these tools in class. This is for you to explore yourself and improve your skills in biological sequence analysis and data visualization. I am just showing you that these are useful tools for various reasons. You can definitely go on your own and start learning how to use these tools for your research projects.</p>
</div>
<div class="section" id="Visualization-of-gene-segments-or-clusters">
<h1>Visualization of gene segments or clusters<a class="headerlink" href="#Visualization-of-gene-segments-or-clusters" title="Permalink to this headline">¶</a></h1>
<p>What about figures like this one shown below, from <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/33115833/">a study</a> to identify a genomic region that might have been horizontally transferred from another organism?</p>
<img alt="gene clusters" src="_images/gene_clusters_ex.jpg" />
<p>As you can see, figures like these shown are very informative as they can display the protein-coding regions in genomes and you can easily see if certain genes are conserved in one or more organisms. They are also useful to display genome rearrangements that may have occurred during the course of their genome evolution.</p>
<p>This figure was created using the tool known as <strong>EasyFig</strong>. The paper describing this tool was published here: <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/21278367/">https://pubmed.ncbi.nlm.nih.gov/21278367/</a></p>
<p>You can download this tool here: <a class="reference external" href="https://sourceforge.net/projects/easyfig/">https://sourceforge.net/projects/easyfig/</a></p>
<p>You can also use other tools such as <a class="reference external" href="http://genoplotr.r-forge.r-project.org/">genoPlotR</a>, which you have used a few weeks ago. <code class="docutils literal notranslate"><span class="pre">genoPlotR</span></code> provides many additional useful features such as being able to display phylogenetic trees to go along with figures of gene clusters or genomic regions. However, it is not an easy tool to learn to use and may require several hours of learning to use its R functions and commands.</p>
<p>Another tool that just came out in recent years to compare and visualize syntenic gene clusters is known as <strong>Clinker</strong>.</p>
<p><a class="reference external" href="https://github.com/gamcil/clinker">https://github.com/gamcil/clinker</a></p>
<p>This tool is written in a completely different programming language: Javascript. It can draw similar figures to <strong>genoPlotR</strong> as you can see here:</p>
<img alt="_images/clinker.png" src="_images/clinker.png" />
<p>Again, be mindful that this tool is designed to display only a short segment of the genomes of two or more very similar organisms. You should not try to align whole genomes and visualize them as it may not work with this tool.</p>
</div>
<div class="section" id="Phylogenetic-trees-visualization">
<h1>Phylogenetic trees visualization<a class="headerlink" href="#Phylogenetic-trees-visualization" title="Permalink to this headline">¶</a></h1>
<p>How about display phylogenetic trees and associated information? You have already learned to use <strong>FigTree</strong>, a graphical tool to display phylogenetic trees and to save output PDF or image files from its display. Although this tool is quite useful, sometimes, you may have hundreds or thousands of taxa that may not display well on FigTree. In those cases, you will sometimes require to use more specialized tree-viewing tool or software.</p>
<p>Take for example, look at this tree below from a paper <a class="reference external" href="https://mbio.asm.org/content/11/1/e02975-19.long">I published early this year</a>.</p>
<img alt="mBio figure" src="_images/mbio_figure.jpg" />
<p>These trees contain thousands of taxa and will not display well on FigTree. For trees like these, you may want to use a tool known as <strong>iTOL</strong>. See here (<a class="reference external" href="https://itol.embl.de/">https://itol.embl.de/</a>) to visit this web-based tool. This tool can display thousands of tree branches as shown and it provides many useful features such as heatmaps, charts, or graphs to go along with the phylogenetic tree. It will also allow you to export the high-quality vector drawing of these tree figures into formats such as SVG, PDF, or
EPS, so that you can use other tools like Adobe Illustrator or Inkscape to touch up and improve the figures for publication.</p>
<p>Sometimes, you need to do more than just displaying phylogenetic trees. You can use phylogenetic/phylogenomic trees to detect horizontal gene transfer events, measure distances between clades of interest (for example, to understand their evolutionary distances), or even perform statistical tests to prove/disprove your hypotheses. For those purposes, you cannot just use a simple phylogenetic displaying tool like FigTree or even iTOL. You need more programmatic tools. Some of these tools are
available in both Python and R languages.</p>
<p>In Python, there is a tool known as <strong>ETE Toolkit</strong> (see here: <a class="reference external" href="http://etetoolkit.org/">http://etetoolkit.org/</a>). It is a Python framework for analysis and visualization of phylogenetic trees. If you visit this page, you will see that the tool can not only display beautiful trees but it can also annotate trees, compare tree topologies, test hypotheses, etc. It provides command-line tools but also a Python library that you can use to develop your own Python scripts to display phylogenetic trees. Python code and tutorial
(with examples) can be found here: <a class="reference external" href="http://etetoolkit.org/docs/latest/tutorial/index.html">http://etetoolkit.org/docs/latest/tutorial/index.html</a></p>
<p>For example, I can use ETE Python library and display a tree within the Jupyter environment. See examples below:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">Tree</span><span class="p">,</span> <span class="n">TreeStyle</span>

<span class="n">t</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span> <span class="s2">&quot;((a,b),c);&quot;</span> <span class="p">)</span>
<span class="n">circular_style</span> <span class="o">=</span> <span class="n">TreeStyle</span><span class="p">()</span>
<span class="n">circular_style</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;c&quot;</span> <span class="c1"># draw tree in circular mode</span>
<span class="n">circular_style</span><span class="o">.</span><span class="n">scale</span> <span class="o">=</span> <span class="mi">20</span>

<span class="n">t</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%%</span><span class="s2">inline&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="_images/exercises_week12_9_0.png" src="_images/exercises_week12_9_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">Tree</span><span class="p">,</span> <span class="n">TreeStyle</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">()</span>
<span class="n">t</span><span class="o">.</span><span class="n">populate</span><span class="p">(</span><span class="mi">30</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">TreeStyle</span><span class="p">()</span>
<span class="n">ts</span><span class="o">.</span><span class="n">show_leaf_name</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">ts</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;c&quot;</span>
<span class="n">ts</span><span class="o">.</span><span class="n">arc_start</span> <span class="o">=</span> <span class="o">-</span><span class="mi">180</span> <span class="c1"># 0 degrees = 3 o&#39;clock</span>
<span class="n">ts</span><span class="o">.</span><span class="n">arc_span</span> <span class="o">=</span> <span class="mi">180</span>
<span class="n">t</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%%</span><span class="s2">inline&quot;</span><span class="p">,</span> <span class="n">tree_style</span><span class="o">=</span><span class="n">ts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="_images/exercises_week12_10_0.png" src="_images/exercises_week12_10_0.png" />
</div>
</div>
<p>As you can see, in just a few simple code, you can display tree structures very easily and without having to mess around with things manually. Tools like ETE Toolkit are very useful if you have thousands of individual trees to display and it would not be feasible to use a graphical tool like FigTree to draw each of the tree figure.</p>
<p>Besides ETE Toolkit, there is another Python-based tool known as <strong>Dendropy</strong> (See here: <a class="reference external" href="https://dendropy.org/">https://dendropy.org/</a>). Dendropy is more of a utility tool for phylogenetic tree analyses rather than for displaying publication-quality tree figures.</p>
<p>In R, there are several tools that allow you to do similar kinds of analyses but also display beautiful tree figures. One of the most well-known R packages is known as <strong>Ape</strong>, which stands for “Analysis of Phylogenetics and Evolution”. You can find it <a class="reference external" href="http://ape-package.ird.fr/">here</a> and <a class="reference external" href="https://cran.r-project.org/web/packages/ape/index.html">here</a>.</p>
<p>Books have been written about this tool and you can see one of the books here:</p>
<p><a class="reference external" href="http://ape-package.ird.fr/APER.html">http://ape-package.ird.fr/APER.html</a></p>
<p>Another, more recent, R package is known as <strong>ggtree</strong>. See here:</p>
<p><a class="reference external" href="https://guangchuangyu.github.io/software/ggtree/">https://guangchuangyu.github.io/software/ggtree/</a></p>
<p>The paper describing ggtree can be found <a class="reference external" href="https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.12628">here</a>. It is similar to ETE Toolkit in terms of the functionality and features it provides, though it is mostly a visualization tool. You can even use it together with other tools such as <code class="docutils literal notranslate"><span class="pre">phyloseq</span></code> you used a few weeks ago to display trees and associated features such as abundances of certain taxa.</p>
<p>For more detailed examples and use cases, see this website:</p>
<p><a class="reference external" href="https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html">https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html</a></p>
</div>
<div class="section" id="Final-comments">
<h1>Final comments<a class="headerlink" href="#Final-comments" title="Permalink to this headline">¶</a></h1>
<p>Again, to reiterate my points made earlier, these are just a few examples of some visualization tools and techniques available and they are by no means comprehensive. Bioinformatic tools are being developed and published every day and some of the tools mentioned here today may become obsolete in a few months. This is the nature of how quickly this field is moving forward. So I would highly encourage you to explore new tools and try to use them yourself and I hope that at least some of the tools
you have learned to use in this course will become useful for your own research in the near future.</p>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
        <p class="logo"><a href="index.html" title="index">
          <img class="logo" src="_static/gm_pan.png" alt="Logo"/>
        </a></p><div class="sphinx-toc sphinxlocaltoc">
    <h3><a href="index.html">Page contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Visualization of microbial genomes</a></li>
<li><a class="reference internal" href="#Visualization-of-gene-segments-or-clusters">Visualization of gene segments or clusters</a></li>
<li><a class="reference internal" href="#Phylogenetic-trees-visualization">Phylogenetic trees visualization</a></li>
<li><a class="reference internal" href="#Final-comments">Final comments</a></li>
</ul>

  </div>
  <div class="sphinxprev">
    <h4>Previous page</h4>
    <p class="topless"><a href="Week12.html"
                          title="Previous page">&larr; Week 12</a></p>
  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/exercises_week12.ipynb.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
    
    
        <div class="sidebar-toggle-group no-js">
            
            <button class="sidebar-toggle" id="sidebar-hide" title="Hide the sidebar menu">
                 «
                <span class="show-for-small">hide menu</span>
                
            </button>
            <button class="sidebar-toggle" id="sidebar-show" title="Show the sidebar menu">
                
                <span class="show-for-small">menu</span>
                <span class="hide-for-small">sidebar</span>
                 »
            </button>
        </div>
    
      <div class="clearer"></div>
    </div>
    <div class="relbar-bottom">
        
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="Week12.html" title="Week 12"
             >previous</a> &nbsp; &nbsp;</li>
    <li><a href="index.html">Microbial Genomics Lab Spring 2022  documentation</a> &#187;</li>

        <li class="nav-item nav-item-this"><a href="">Visualization of microbial genomes</a></li> 
      </ul>
    </div>
    </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2022, Jimmy Saw.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.3.2.
    </div>
    <!-- cloud_sptheme 1.4 -->
  </body>
</html>