


<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Genome Assembly using SPAdes &#8212; Microbial Genomics Lab Spring 2022  documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/cloud.css" />
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Noticia+Text:400,i,b,bi|Open+Sans:400,i,b,bi|Roboto+Mono:400,i,b,bi&amp;display=swap" type="text/css" />
    
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>

    
    
     
        <script src="_static/jquery.cookie.js"></script>
    

    
     
        <script src="_static/cloud.base.js"></script>
    

    
     
        <script src="_static/cloud.js"></script>
    

    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
        <meta name="viewport" content="width=device-width, initial-scale=1">
  </head><body>
    <div class="relbar-top">
        
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
    <li><a href="index.html">Microbial Genomics Lab Spring 2022  documentation</a> &#187;</li>

        <li class="nav-item nav-item-this"><a href="">Genome Assembly using SPAdes</a></li> 
      </ul>
    </div>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for cloud theme */

/* nicer titles and more space for info and warning logos */

div.admonition p.admonition-title {
    background: rgba(0, 0, 0, .05);
    margin: .5em -1em;
    margin-top: -.5em !important;
    padding: .5em .5em .5em 2.65em;
}

/* indent single paragraph */
div.admonition {
    text-indent: 20px;
}
/* don't indent multiple paragraphs */
div.admonition > p {
    text-indent: 0;
}
/* remove excessive padding */
div.admonition.inline-title p.admonition-title {
    padding-left: .2em;
}
</style>
<div class="section" id="Genome-Assembly-using-SPAdes">
<h1>Genome Assembly using SPAdes<a class="headerlink" href="#Genome-Assembly-using-SPAdes" title="Permalink to this headline">¶</a></h1>
<div class="section" id="Introduction-to-SPAdes">
<h2>Introduction to SPAdes<a class="headerlink" href="#Introduction-to-SPAdes" title="Permalink to this headline">¶</a></h2>
<p>SPAdes was originally designed as a tool to assemble single-cell genomic sequences that usualy comprise uneven sequence coverage. In order to be able to discern such unenven coverages, the authors developed special algorithms to detect such coverages and to account for such regions. Also, to correct for errors that might be introduced during multiple-displacement amplification (MDA), SPAdes incorporates an “error-correction” step to improve sequence quality in those regions.</p>
<p>Since its intial development in 2012, the assembler has evolved to be able to handle different types of sequencing technologies and datasets. Now, it can handle not only single-cell genomic data but also genomic sequences from isolate genomes (cultures), metagenomes. It can also handle latest “next-gen sequencing” (NGS) data such as PacBio, Nanopore, etc. SPAdes is hosted at this website (<a class="reference external" href="https://cab.spbu.ru/software/spades/">https://cab.spbu.ru/software/spades/</a>) but you can also visit their github page to see the latest version
here (<a class="reference external" href="https://github.com/ablab/spades">https://github.com/ablab/spades</a>). The github page contains most updated information on how to run SPAdes and what options you can use.</p>
<p>In today’s computational exercise, you will use SPAdes to assemble a microbial genome sequenced with Illumina sequencing technology. In order to be able to run SPAdes, you will need to install SPAdes through the Miniconda Python package managing system.</p>
</div>
<div class="section" id="Installing-SPAdes">
<h2>Installing SPAdes<a class="headerlink" href="#Installing-SPAdes" title="Permalink to this headline">¶</a></h2>
<p>To install SPAdes on your laptop, you will first type this following command:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">search</span> <span class="n">spades</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Loading channels: done
# Name                       Version           Build  Channel
spades                         3.6.2               0  bioconda
spades                         3.7.0               0  bioconda
spades                         3.8.0               0  bioconda
spades                         3.8.1               0  bioconda
spades                         3.9.0               0  bioconda
spades                         3.9.0          py27_1  bioconda
spades                         3.9.0          py27_2  bioconda
spades                         3.9.0          py34_1  bioconda
spades                         3.9.0          py35_1  bioconda
spades                         3.9.0          py35_2  bioconda
spades                         3.9.1               0  bioconda
spades                        3.10.0          py27_0  bioconda
spades                        3.10.0          py34_0  bioconda
spades                        3.10.0          py35_0  bioconda
spades                        3.10.1               1  bioconda
spades                        3.10.1          py27_0  bioconda
spades                        3.10.1          py34_0  bioconda
spades                        3.10.1          py35_0  bioconda
spades                        3.11.0          py27_0  bioconda
spades                        3.11.0          py27_1  bioconda
spades                        3.11.0          py35_0  bioconda
spades                        3.11.0          py35_1  bioconda
spades                        3.11.0          py36_0  bioconda
spades                        3.11.0          py36_1  bioconda
spades                        3.11.1      h21aa3a5_2  bioconda
spades                        3.11.1      h21aa3a5_3  bioconda
spades                        3.11.1      hb249e0d_4  bioconda
spades                        3.11.1      he4cf2ce_2  bioconda
spades                        3.11.1          py27_0  bioconda
spades                        3.11.1          py27_1  bioconda
spades                        3.11.1          py35_0  bioconda
spades                        3.11.1          py35_1  bioconda
spades                        3.11.1          py36_0  bioconda
spades                        3.11.1          py36_1  bioconda
spades                        3.12.0               1  bioconda
spades                        3.12.0          py27_0  bioconda
spades                        3.12.0          py35_0  bioconda
spades                        3.12.0          py36_0  bioconda
spades                        3.13.0               0  bioconda
spades                        3.13.1               0  bioconda
spades                        3.13.1      h21ccd2c_2  bioconda
spades                        3.13.1      h8f8a155_1  bioconda
spades                        3.13.2      h21ccd2c_0  bioconda
spades                        3.14.0      h21ccd2c_0  bioconda
spades                        3.14.1      he641558_0  bioconda

Note: you may need to restart the kernel to use updated packages.
</pre></div></div>
</div>
<p>This brings up all the different versions of SPAdes currently on my <code class="docutils literal notranslate"><span class="pre">bioconda</span></code> channel. The latest version tends to be at the bottom of the page. Then to install the latest version, you would just type <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">spades</span></code>. On my laptop, I have already installed a version of SPAdes. To check what version I have already install, I can type like this below:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="nb">list</span> <span class="o">|</span> <span class="n">grep</span> <span class="n">spades</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
spades                    3.13.0                        0    bioconda

Note: you may need to restart the kernel to use updated packages.
</pre></div></div>
</div>
<p>So, you can see that I have version 3.13.0 installed. If I want to be explicit about the specific version of SPAdes I want to install, I can type like this:</p>
<p><code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">spades=3.14.0</span></code></p>
<p>This will install version 3.14.0. If you do not specify a version number, then it will install the latest version by default. Once you have SPAdes installed, you should be able to type this command <code class="docutils literal notranslate"><span class="pre">spades.py</span></code> right away in your terminal.</p>
</div>
<div class="section" id="Running-SPAdes">
<h2>Running SPAdes<a class="headerlink" href="#Running-SPAdes" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
spades.py
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
SPAdes genome assembler v3.13.0

Usage: /Users/jsaw/miniconda3/bin/spades.py [options] -o &lt;output_dir&gt;

Basic options:
-o      &lt;output_dir&gt;      directory to store all the resulting files (required)
--sc                    this flag is required for MDA (single-cell) data
--meta                  this flag is required for metagenomic sample data
--rna                   this flag is required for RNA-Seq data
--plasmid               runs plasmidSPAdes pipeline for plasmid detection
--iontorrent            this flag is required for IonTorrent data
--test                  runs SPAdes on toy dataset
-h/--help               prints this usage message
-v/--version            prints version

Input data:
--12    &lt;filename&gt;        file with interlaced forward and reverse paired-end reads
-1      &lt;filename&gt;        file with forward paired-end reads
-2      &lt;filename&gt;        file with reverse paired-end reads
-s      &lt;filename&gt;        file with unpaired reads
--merged        &lt;filename&gt;        file with merged forward and reverse paired-end reads
--pe&lt;#&gt;-12        &lt;filename&gt;        file with interlaced reads for paired-end library number &lt;#&gt; (&lt;#&gt; = 1,2,...,9)
--pe&lt;#&gt;-1 &lt;filename&gt;        file with forward reads for paired-end library number &lt;#&gt; (&lt;#&gt; = 1,2,...,9)
--pe&lt;#&gt;-2 &lt;filename&gt;        file with reverse reads for paired-end library number &lt;#&gt; (&lt;#&gt; = 1,2,...,9)
--pe&lt;#&gt;-s &lt;filename&gt;        file with unpaired reads for paired-end library number &lt;#&gt; (&lt;#&gt; = 1,2,...,9)
--pe&lt;#&gt;-m &lt;filename&gt;        file with merged reads for paired-end library number &lt;#&gt; (&lt;#&gt; = 1,2,...,9)
--pe&lt;#&gt;-&lt;or&gt;        orientation of reads for paired-end library number &lt;#&gt; (&lt;#&gt; = 1,2,...,9; &lt;or&gt; = fr, rf, ff)
--s&lt;#&gt;            &lt;filename&gt;        file with unpaired reads for single reads library number &lt;#&gt; (&lt;#&gt; = 1,2,...,9)
--mp&lt;#&gt;-12        &lt;filename&gt;        file with interlaced reads for mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--mp&lt;#&gt;-1 &lt;filename&gt;        file with forward reads for mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--mp&lt;#&gt;-2 &lt;filename&gt;        file with reverse reads for mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--mp&lt;#&gt;-s &lt;filename&gt;        file with unpaired reads for mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--mp&lt;#&gt;-&lt;or&gt;        orientation of reads for mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9; &lt;or&gt; = fr, rf, ff)
--hqmp&lt;#&gt;-12      &lt;filename&gt;        file with interlaced reads for high-quality mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--hqmp&lt;#&gt;-1       &lt;filename&gt;        file with forward reads for high-quality mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--hqmp&lt;#&gt;-2       &lt;filename&gt;        file with reverse reads for high-quality mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--hqmp&lt;#&gt;-s       &lt;filename&gt;        file with unpaired reads for high-quality mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--hqmp&lt;#&gt;-&lt;or&gt;      orientation of reads for high-quality mate-pair library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9; &lt;or&gt; = fr, rf, ff)
--nxmate&lt;#&gt;-1     &lt;filename&gt;        file with forward reads for Lucigen NxMate library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--nxmate&lt;#&gt;-2     &lt;filename&gt;        file with reverse reads for Lucigen NxMate library number &lt;#&gt; (&lt;#&gt; = 1,2,..,9)
--sanger        &lt;filename&gt;        file with Sanger reads
--pacbio        &lt;filename&gt;        file with PacBio reads
--nanopore      &lt;filename&gt;        file with Nanopore reads
--tslr  &lt;filename&gt;        file with TSLR-contigs
--trusted-contigs       &lt;filename&gt;        file with trusted contigs
--untrusted-contigs     &lt;filename&gt;        file with untrusted contigs

Pipeline options:
--only-error-correction runs only read error correction (without assembling)
--only-assembler        runs only assembling (without read error correction)
--careful               tries to reduce number of mismatches and short indels
--continue              continue run from the last available check-point
--restart-from  &lt;cp&gt;      restart run with updated options and from the specified check-point (&#39;ec&#39;, &#39;as&#39;, &#39;k&lt;int&gt;&#39;, &#39;mc&#39;, &#39;last&#39;)
--disable-gzip-output   forces error correction not to compress the corrected reads
--disable-rr            disables repeat resolution stage of assembling

Advanced options:
--dataset       &lt;filename&gt;        file with dataset description in YAML format
-t/--threads    &lt;int&gt;             number of threads
                                [default: 16]
-m/--memory     &lt;int&gt;             RAM limit for SPAdes in Gb (terminates if exceeded)
                                [default: 250]
--tmp-dir       &lt;dirname&gt; directory for temporary files
                                [default: &lt;output_dir&gt;/tmp]
-k              &lt;int,int,...&gt;     comma-separated list of k-mer sizes (must be odd and
                                less than 128) [default: &#39;auto&#39;]
--cov-cutoff    &lt;float&gt;           coverage cutoff value (a positive float number, or &#39;auto&#39;, or &#39;off&#39;) [default: &#39;off&#39;]
--phred-offset  &lt;33 or 64&gt;        PHRED quality offset in the input reads (33 or 64)
                                [default: auto-detect]
</pre></div></div>
</div>
<p>I have to type <code class="docutils literal notranslate"><span class="pre">%%bash</span></code> before my command in order to be able to show command line outputs in Jupyter Lab environment. When you are typing <code class="docutils literal notranslate"><span class="pre">spades.py</span></code> in your own terminal, you do not need to type this <code class="docutils literal notranslate"><span class="pre">%%bash</span></code> command. As you can see, you are given a number of options that you can specify when running SPAdes. For example, if you are running a metagenome assembly using SPAdes, then you would add an <code class="docutils literal notranslate"><span class="pre">--meta</span></code> flag to indicate the program that you are running a metagenome assembly.</p>
<p>In today’s exercise, you will run SPAdes to assemble a genome belonging to a bacterium (<em>Salmonella enterica</em>) that you have downloaded previously from the following website: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/sra/SRX9094324">https://www.ncbi.nlm.nih.gov/sra/SRX9094324</a></p>
<p>You have also done processing of this raw data into a form that is suitable to be used as an input for the SPAdes assembler.</p>
</div>
<div class="section" id="Checking-sequence-qualities">
<h2>Checking sequence qualities<a class="headerlink" href="#Checking-sequence-qualities" title="Permalink to this headline">¶</a></h2>
<p>To check sequence qualities of the two <code class="docutils literal notranslate"><span class="pre">fastq</span></code> files you have downloaded previously, you will run the <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> tool. First, using your terminal, nagivate to the folder where you have downloaded these sequences. For example, if you downloaded the sequences in <code class="docutils literal notranslate"><span class="pre">/home/yourname/data</span></code>, then you would type:</p>
<p><code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">/home/yourname/data</span></code></p>
<p>In my case, I have downloaded them into another folder named <code class="docutils literal notranslate"><span class="pre">data</span></code>.</p>
<p>To check the quality of the downloaded sequences, you would run the FastQC tool, which you have installed previously. To run, navigate into the folder containing the data, then type:</p>
<p><code class="docutils literal notranslate"><span class="pre">fastqc</span> <span class="pre">*.fastq</span></code></p>
<p>Inspect the quality and statistics reported in this html file to see what the report is telling you. Remember, these are paired fastq files and this means “SRR12610971_1_fastqc.html” is only one of the pair. You will need to inspect the second file “SRR12610971_2_fastqc.html” as well.</p>
</div>
<div class="section" id="Trimming-the-sequences">
<h2>Trimming the sequences<a class="headerlink" href="#Trimming-the-sequences" title="Permalink to this headline">¶</a></h2>
<p>To trim the possible Illumina adapter sequences and poor quality regions, you can use several trimming tools such as Trimmomatic, bbduk, or cutadapt. Here, you will use bbduk, which is part of a suite of tools that come with bbmap tool. You can run bbduk with the following commands:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bbduk.sh -Xmx12g <span class="nv">ktrim</span><span class="o">=</span>r ordered <span class="nv">minlen</span><span class="o">=</span><span class="m">50</span> <span class="nv">mink</span><span class="o">=</span><span class="m">11</span> <span class="nv">tbo</span><span class="o">=</span>t <span class="nv">rcomp</span><span class="o">=</span>f <span class="nv">k</span><span class="o">=</span><span class="m">21</span> <span class="nv">ow</span><span class="o">=</span>t <span class="nv">zl</span><span class="o">=</span><span class="m">4</span> <span class="se">\</span>
        <span class="nv">qtrim</span><span class="o">=</span>rl <span class="nv">trimq</span><span class="o">=</span><span class="m">20</span> <span class="se">\</span>
        <span class="nv">in1</span><span class="o">=</span>SRR12610971_1.fastq.gz <span class="se">\</span>
        <span class="nv">in2</span><span class="o">=</span>SRR12610971_2.fastq.gz <span class="se">\</span>
        <span class="nv">ref</span><span class="o">=</span>~/databases/adapters/adapters.fa  <span class="se">\</span>
        <span class="nv">out1</span><span class="o">=</span>SRR12610971_1.trimmed.fastq.gz <span class="se">\</span>
        <span class="nv">out2</span><span class="o">=</span>SRR12610971_2.trimmed.fastq.gz
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">-Xmx12g</span></code> parameter I am providing here is to tell <code class="docutils literal notranslate"><span class="pre">bbduk.sh</span></code> that I want to use 12GB of my computer’s memory for this task. This parameter may look a bit different from the one shown a few weeks ago.</p>
<p>It will take a few minutes to complete this task. Note that I am using the “\” to enter the command in multiple lines due to the long list of parameters I need to specify with this tool. You can actually type the whole thing in one line without having to type the “\” but it is more difficult to see all the options in one screen. There are several parameters that I am entering here to tell the trimming tool what to do exactly with the raw sequences. Detailed options/parameters available with
this tool can be seen when you type this:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
bbduk.sh -h
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Written by Brian Bushnell
Last modified March 24, 2020

Description:  Compares reads to the kmers in a reference dataset, optionally
allowing an edit distance. Splits the reads into two outputs - those that
match the reference, and those that don&#39;t. Can also trim (remove) the matching
parts of the reads rather than binning the reads.
Please read bbmap/docs/guides/BBDukGuide.txt for more information.

Usage:  bbduk.sh in=&lt;input file&gt; out=&lt;output file&gt; ref=&lt;contaminant files&gt;

Input may be stdin or a fasta or fastq file, compressed or uncompressed.
If you pipe via stdin/stdout, please include the file type; e.g. for gzipped
fasta input, set in=stdin.fa.gz

Input parameters:
in=&lt;file&gt;           Main input. in=stdin.fq will pipe from stdin.
in2=&lt;file&gt;          Input for 2nd read of pairs in a different file.
ref=&lt;file,file&gt;     Comma-delimited list of reference files.
                    In addition to filenames, you may also use the keywords:
                    adapters, artifacts, phix, lambda, pjet, mtst, kapa
literal=&lt;seq,seq&gt;   Comma-delimited list of literal reference sequences.
touppercase=f       (tuc) Change all bases upper-case.
interleaved=auto    (int) t/f overrides interleaved autodetection.
qin=auto            Input quality offset: 33 (Sanger), 64, or auto.
reads=-1            If positive, quit after processing X reads or pairs.
copyundefined=f     (cu) Process non-AGCT IUPAC reference bases by making all
                    possible unambiguous copies.  Intended for short motifs
                    or adapter barcodes, as time/memory use is exponential.
samplerate=1        Set lower to only process a fraction of input reads.
samref=&lt;file&gt;       Optional reference fasta for processing sam files.

Output parameters:
out=&lt;file&gt;          (outnonmatch) Write reads here that do not contain
                    kmers matching the database.  &#39;out=stdout.fq&#39; will pipe
                    to standard out.
out2=&lt;file&gt;         (outnonmatch2) Use this to write 2nd read of pairs to a
                    different file.
outm=&lt;file&gt;         (outmatch) Write reads here that fail filters.  In default
                    kfilter mode, this means any read with a matching kmer.
                    In any mode, it also includes reads that fail filters such
                    as minlength, mingc, maxgc, entropy, etc.  In other words,
                    it includes all reads that do not go to &#39;out&#39;.
outm2=&lt;file&gt;        (outmatch2) Use this to write 2nd read of pairs to a
                    different file.
outs=&lt;file&gt;         (outsingle) Use this to write singleton reads whose mate
                    was trimmed shorter than minlen.
stats=&lt;file&gt;        Write statistics about which contamininants were detected.
refstats=&lt;file&gt;     Write statistics on a per-reference-file basis.
rpkm=&lt;file&gt;         Write RPKM for each reference sequence (for RNA-seq).
dump=&lt;file&gt;         Dump kmer tables to a file, in fasta format.
duk=&lt;file&gt;          Write statistics in duk&#39;s format. *DEPRECATED*
nzo=t               Only write statistics about ref sequences with nonzero hits.
overwrite=t         (ow) Grant permission to overwrite files.
showspeed=t         (ss) &#39;f&#39; suppresses display of processing speed.
ziplevel=2          (zl) Compression level; 1 (min) through 9 (max).
fastawrap=70        Length of lines in fasta output.
qout=auto           Output quality offset: 33 (Sanger), 64, or auto.
statscolumns=3      (cols) Number of columns for stats output, 3 or 5.
                    5 includes base counts.
rename=f            Rename reads to indicate which sequences they matched.
refnames=f          Use names of reference files rather than scaffold IDs.
trd=f               Truncate read and ref names at the first whitespace.
ordered=f           Set to true to output reads in same order as input.
maxbasesout=-1      If positive, quit after writing approximately this many
                    bases to out (outu/outnonmatch).
maxbasesoutm=-1     If positive, quit after writing approximately this many
                    bases to outm (outmatch).
json=f              Print to screen in json format.

Histogram output parameters:
bhist=&lt;file&gt;        Base composition histogram by position.
qhist=&lt;file&gt;        Quality histogram by position.
qchist=&lt;file&gt;       Count of bases with each quality value.
aqhist=&lt;file&gt;       Histogram of average read quality.
bqhist=&lt;file&gt;       Quality histogram designed for box plots.
lhist=&lt;file&gt;        Read length histogram.
phist=&lt;file&gt;        Polymer length histogram.
gchist=&lt;file&gt;       Read GC content histogram.
enthist=&lt;file&gt;      Read entropy histogram.
ihist=&lt;file&gt;        Insert size histogram, for paired reads in mapped sam.
gcbins=100          Number gchist bins.  Set to &#39;auto&#39; to use read length.
maxhistlen=6000     Set an upper bound for histogram lengths; higher uses
                    more memory.  The default is 6000 for some histograms
                    and 80000 for others.

Histograms for mapped sam/bam files only:
histbefore=t        Calculate histograms from reads before processing.
ehist=&lt;file&gt;        Errors-per-read histogram.
qahist=&lt;file&gt;       Quality accuracy histogram of error rates versus quality
                    score.
indelhist=&lt;file&gt;    Indel length histogram.
mhist=&lt;file&gt;        Histogram of match, sub, del, and ins rates by position.
idhist=&lt;file&gt;       Histogram of read count versus percent identity.
idbins=100          Number idhist bins.  Set to &#39;auto&#39; to use read length.
varfile=&lt;file&gt;      Ignore substitution errors listed in this file when
                    calculating error rates.  Can be generated with
                    CallVariants.
vcf=&lt;file&gt;          Ignore substitution errors listed in this VCF file
                    when calculating error rates.
ignorevcfindels=t   Also ignore indels listed in the VCF.

Processing parameters:
k=27                Kmer length used for finding contaminants.  Contaminants
                    shorter than k will not be found.  k must be at least 1.
rcomp=t             Look for reverse-complements of kmers in addition to
                    forward kmers.
maskmiddle=t        (mm) Treat the middle base of a kmer as a wildcard, to
                    increase sensitivity in the presence of errors.
minkmerhits=1       (mkh) Reads need at least this many matching kmers
                    to be considered as matching the reference.
minkmerfraction=0.0 (mkf) A reads needs at least this fraction of its total
                    kmers to hit a ref, in order to be considered a match.
                    If this and minkmerhits are set, the greater is used.
mincovfraction=0.0  (mcf) A reads needs at least this fraction of its total
                    bases to be covered by ref kmers to be considered a match.
                    If specified, mcf overrides mkh and mkf.
hammingdistance=0   (hdist) Maximum Hamming distance for ref kmers (subs only).
                    Memory use is proportional to (3*K)^hdist.
qhdist=0            Hamming distance for query kmers; impacts speed, not memory.
editdistance=0      (edist) Maximum edit distance from ref kmers (subs
                    and indels).  Memory use is proportional to (8*K)^edist.
hammingdistance2=0  (hdist2) Sets hdist for short kmers, when using mink.
qhdist2=0           Sets qhdist for short kmers, when using mink.
editdistance2=0     (edist2) Sets edist for short kmers, when using mink.
forbidn=f           (fn) Forbids matching of read kmers containing N.
                    By default, these will match a reference &#39;A&#39; if
                    hdist&gt;0 or edist&gt;0, to increase sensitivity.
removeifeitherbad=t (rieb) Paired reads get sent to &#39;outmatch&#39; if either is
                    match (or either is trimmed shorter than minlen).
                    Set to false to require both.
trimfailures=f      Instead of discarding failed reads, trim them to 1bp.
                    This makes the statistics a bit odd.
findbestmatch=f     (fbm) If multiple matches, associate read with sequence
                    sharing most kmers.  Reduces speed.
skipr1=f            Don&#39;t do kmer-based operations on read 1.
skipr2=f            Don&#39;t do kmer-based operations on read 2.
ecco=f              For overlapping paired reads only.  Performs error-
                    correction with BBMerge prior to kmer operations.
recalibrate=f       (recal) Recalibrate quality scores.  Requires calibration
                    matrices generated by CalcTrueQuality.
sam=&lt;file,file&gt;     If recalibration is desired, and matrices have not already
                    been generated, BBDuk will create them from the sam file.
amino=f             Run in amino acid mode.  Some features have not been
                    tested, but kmer-matching works fine.  Maximum k is 12.

Speed and Memory parameters:
threads=auto        (t) Set number of threads to use; default is number of
                    logical processors.
prealloc=f          Preallocate memory in table.  Allows faster table loading
                    and more efficient memory usage, for a large reference.
monitor=f           Kill this process if it crashes.  monitor=600,0.01 would
                    kill after 600 seconds under 1% usage.
minrskip=1          (mns) Force minimal skip interval when indexing reference
                    kmers.  1 means use all, 2 means use every other kmer, etc.
maxrskip=1          (mxs) Restrict maximal skip interval when indexing
                    reference kmers. Normally all are used for scaffolds&lt;100kb,
                    but with longer scaffolds, up to maxrskip-1 are skipped.
rskip=              Set both minrskip and maxrskip to the same value.
                    If not set, rskip will vary based on sequence length.
qskip=1             Skip query kmers to increase speed.  1 means use all.
speed=0             Ignore this fraction of kmer space (0-15 out of 16) in both
                    reads and reference.  Increases speed and reduces memory.
Note: Do not use more than one of &#39;speed&#39;, &#39;qskip&#39;, and &#39;rskip&#39;.

Trimming/Filtering/Masking parameters:
Note - if ktrim, kmask, and ksplit are unset, the default behavior is kfilter.
All kmer processing modes are mutually exclusive.
Reads only get sent to &#39;outm&#39; purely based on kmer matches in kfilter mode.

ktrim=f             Trim reads to remove bases matching reference kmers.
                    Values:
                       f (don&#39;t trim),
                       r (trim to the right),
                       l (trim to the left)
kmask=              Replace bases matching ref kmers with another symbol.
                    Allows any non-whitespace character, and processes short
                    kmers on both ends if mink is set.  &#39;kmask=lc&#39; will
                    convert masked bases to lowercase.
maskfullycovered=f  (mfc) Only mask bases that are fully covered by kmers.
ksplit=f            For single-ended reads only.  Reads will be split into
                    pairs around the kmer.  If the kmer is at the end of the
                    read, it will be trimmed instead.  Singletons will go to
                    out, and pairs will go to outm.  Do not use ksplit with
                    other operations such as quality-trimming or filtering.
mink=0              Look for shorter kmers at read tips down to this length,
                    when k-trimming or masking.  0 means disabled.  Enabling
                    this will disable maskmiddle.
qtrim=f             Trim read ends to remove bases with quality below trimq.
                    Performed AFTER looking for kmers.  Values:
                       rl (trim both ends),
                       f (neither end),
                       r (right end only),
                       l (left end only),
                       w (sliding window).
trimq=6             Regions with average quality BELOW this will be trimmed,
                    if qtrim is set to something other than f.  Can be a
                    floating-point number like 7.3.
trimclip=f          Trim soft-clipped bases from sam files.
minlength=10        (ml) Reads shorter than this after trimming will be
                    discarded.  Pairs will be discarded if both are shorter.
mlf=0               (minlengthfraction) Reads shorter than this fraction of
                    original length after trimming will be discarded.
maxlength=          Reads longer than this after trimming will be discarded.
minavgquality=0     (maq) Reads with average quality (after trimming) below
                    this will be discarded.
maqb=0              If positive, calculate maq from this many initial bases.
minbasequality=0    (mbq) Reads with any base below this quality (after
                    trimming) will be discarded.
maxns=-1            If non-negative, reads with more Ns than this
                    (after trimming) will be discarded.
mcb=0               (minconsecutivebases) Discard reads without at least
                    this many consecutive called bases.
ottm=f              (outputtrimmedtomatch) Output reads trimmed to shorter
                    than minlength to outm rather than discarding.
tp=0                (trimpad) Trim this much extra around matching kmers.
tbo=f               (trimbyoverlap) Trim adapters based on where paired
                    reads overlap.
strictoverlap=t     Adjust sensitivity for trimbyoverlap mode.
minoverlap=14       Require this many bases of overlap for detection.
mininsert=40        Require insert size of at least this for overlap.
                    Should be reduced to 16 for small RNA sequencing.
tpe=f               (trimpairsevenly) When kmer right-trimming, trim both
                    reads to the minimum length of either.
forcetrimleft=0     (ftl) If positive, trim bases to the left of this position
                    (exclusive, 0-based).
forcetrimright=0    (ftr) If positive, trim bases to the right of this position
                    (exclusive, 0-based).
forcetrimright2=0   (ftr2) If positive, trim this many bases on the right end.
forcetrimmod=0      (ftm) If positive, right-trim length to be equal to zero,
                    modulo this number.
restrictleft=0      If positive, only look for kmer matches in the
                    leftmost X bases.
restrictright=0     If positive, only look for kmer matches in the
                    rightmost X bases.
mingc=0             Discard reads with GC content below this.
maxgc=1             Discard reads with GC content above this.
gcpairs=t           Use average GC of paired reads.
                    Also affects gchist.
tossjunk=f          Discard reads with invalid characters as bases.
swift=f             Trim Swift sequences: Trailing C/T/N R1, leading G/A/N R2.

Header-parsing parameters - these require Illumina headers:
chastityfilter=f    (cf) Discard reads with id containing &#39; 1:Y:&#39; or &#39; 2:Y:&#39;.
barcodefilter=f     Remove reads with unexpected barcodes if barcodes is set,
                    or barcodes containing &#39;N&#39; otherwise.  A barcode must be
                    the last part of the read header.  Values:
                       t:     Remove reads with bad barcodes.
                       f:     Ignore barcodes.
                       crash: Crash upon encountering bad barcodes.
barcodes=           Comma-delimited list of barcodes or files of barcodes.
xmin=-1             If positive, discard reads with a lesser X coordinate.
ymin=-1             If positive, discard reads with a lesser Y coordinate.
xmax=-1             If positive, discard reads with a greater X coordinate.
ymax=-1             If positive, discard reads with a greater Y coordinate.

Polymer trimming:
trimpolya=0         If greater than 0, trim poly-A or poly-T tails of
                    at least this length on either end of reads.
trimpolygleft=0     If greater than 0, trim poly-G prefixes of at least this
                    length on the left end of reads.  Does not trim poly-C.
trimpolygright=0    If greater than 0, trim poly-G tails of at least this
                    length on the right end of reads.  Does not trim poly-C.
trimpolyg=0         This sets both left and right at once.
filterpolyg=0       If greater than 0, remove reads with a poly-G prefix of
                    at least this length (on the left).
Note: there are also equivalent poly-C flags.

Polymer tracking:
pratio=base,base    &#39;pratio=G,C&#39; will print the ratio of G to C polymers.
plen=20             Length of homopolymers to count.

Entropy/Complexity parameters:
entropy=-1          Set between 0 and 1 to filter reads with entropy below
                    that value.  Higher is more stringent.
entropywindow=50    Calculate entropy using a sliding window of this length.
entropyk=5          Calculate entropy using kmers of this length.
minbasefrequency=0  Discard reads with a minimum base frequency below this.
entropytrim=f       Values:
                       f:  (false) Do not entropy-trim.
                       r:  (right) Trim low entropy on the right end only.
                       l:  (left) Trim low entropy on the left end only.
                       rl: (both) Trim low entropy on both ends.
entropymask=f       Values:
                       f:  (filter) Discard low-entropy sequences.
                       t:  (true) Mask low-entropy parts of sequences with N.
                       lc: Change low-entropy parts of sequences to lowercase.
entropymark=f       Mark each base with its entropy value.  This is on a scale
                    of 0-41 and is reported as quality scores, so the output
                    should be fastq or fasta+qual.
NOTE: If set, entropytrim overrides entropymask.

Cardinality estimation:
cardinality=f       (loglog) Count unique kmers using the LogLog algorithm.
cardinalityout=f    (loglogout) Count unique kmers in output reads.
loglogk=31          Use this kmer length for counting.
loglogbuckets=2048  Use this many buckets for counting.
khist=&lt;file&gt;        Kmer frequency histogram; plots number of kmers versus
                    kmer depth.  This is approximate.
khistout=&lt;file&gt;     Kmer frequency histogram for output reads.

Java Parameters:

-Xmx                This will set Java&#39;s memory usage, overriding autodetection.
                    -Xmx20g will
                    specify 20 gigs of RAM, and -Xmx200m will specify 200 megs.
                    The max is typically 85% of physical memory.
-eoom               This flag will cause the process to exit if an
                    out-of-memory exception occurs.  Requires Java 8u92+.
-da                 Disable assertions.

Please contact Brian Bushnell at bbushnell@lbl.gov if you encounter any problems.

</pre></div></div>
</div>
<p>As you can see, there are several options and what each of these options does when you select a certain one. In some cases, you can specify only one option but not together with another one.</p>
</div>
<div class="section" id="Run-fastqc-on-quality-trimmed-sequences-to-compare-its-quality-with-untrimmed-sequences">
<h2>Run fastqc on quality-trimmed sequences to compare its quality with untrimmed sequences<a class="headerlink" href="#Run-fastqc-on-quality-trimmed-sequences-to-compare-its-quality-with-untrimmed-sequences" title="Permalink to this headline">¶</a></h2>
<p>Now that you have done trimming the raw sequences, you will then check what the quality looks like. Run fastqc again on trimmed sequences by typing <code class="docutils literal notranslate"><span class="pre">fastqc</span> <span class="pre">*.trimmed.fastq.gz</span></code> in the folder where the trimmed files are located. You should inspect the quality reports produced by fastqc to make sure it did what you intended it to do.</p>
<p>You can now use the trimmed fastq files as input to run SPAdes.</p>
</div>
<div class="section" id="Running-the-assembly-on-your-laptop-or-home-computer">
<h2>Running the assembly on your laptop or home computer<a class="headerlink" href="#Running-the-assembly-on-your-laptop-or-home-computer" title="Permalink to this headline">¶</a></h2>
<p>SPAdes can be run on your home desktop or a laptop computer with a given example dataset provided. It will take a while to complete but it will get done. To run, type the following commands (shown as an example):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>spades.py <span class="se">\</span>
    -t <span class="m">4</span> <span class="se">\</span>
    --pe1-1 data/SRR12610971_1.trimmed.fastq.gz <span class="se">\</span>
    --pe1-2 data/SRR12610971_2.trimmed.fastq.gz <span class="se">\</span>
    -k <span class="m">21</span>,33,55,77 <span class="se">\</span>
    -o spades_assembly
</pre></div>
</div>
<p>This example is one of the simplest examples to run a genome assembly using this particular assembler. Again, this tool also has a number of options/parameters that you can specify, depending on the type of data you have, the size of the data, etc. In this example, <code class="docutils literal notranslate"><span class="pre">-t</span> <span class="pre">4</span></code> means I am using 4 cpus to run this tool. If you have more cpu cores on your computer, you can specify a larger number. However, with larger datasets (such as metagenomic data), you will require not only more cpus but also
larger amount of memory. For this, you will require the use of a high-performance computaional (HPC) clusters that will have larger memory and cpu cores.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">-k</span> <span class="pre">21,33,55,77</span></code> means that I am using 4 different k-mers (read the original SPAdes paper) to run this assembly. You can specify more k-mers but this may cause the assembler to run longer and may not necessarily lead to improved assemblies. This is more of an optimization problem. By design, k-mers need to be odd numbers.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">spades_assembly</span></code> parameter tells the program to put output files into a new directory named ‘spades_assembly’. To see more options, type:</p>
<p><code class="docutils literal notranslate"><span class="pre">spades.py</span> <span class="pre">-h</span></code></p>
<p>This particular data set run on my iMac with 4 cpus took less than 1 hour to complete. Let it run until it finishes. Next task for you would be to check the assembly quality after it is done.</p>
</div>
<div class="section" id="Running-the-assembly-on-a-cluster-at-GW">
<h2>Running the assembly on a cluster at GW<a class="headerlink" href="#Running-the-assembly-on-a-cluster-at-GW" title="Permalink to this headline">¶</a></h2>
<p>While the previous example of SPAdes assembly using <em>Salmonella entrica</em> sequences you have downloaded from NCBI can be done on your laptop, larger datasets such as metagenomes you have downloaded during the 2nd week of the class are too big to be assembled on your laptop. They will need to be run on much more powerful computers such as high-performance computational (HPC) clusters we have at GW.</p>
<p>We will be connceting to a remote HPC cluster known as <code class="docutils literal notranslate"><span class="pre">cerberus</span></code> next week to perform genome or metagenome assemblies of larger datasets.</p>
</div>
<div class="section" id="Checking-the-quality-of-assemblies">
<h2>Checking the quality of assemblies<a class="headerlink" href="#Checking-the-quality-of-assemblies" title="Permalink to this headline">¶</a></h2>
<p>Next thing to do after running these assemblies is to assess how well the assembly was and some other basic metrics related to the assembly. Remember, you started with millions of individual DNA fragments which are only a few hundred bases long and the assembler has magically put together these highly complicated puzzle pieces into a more complete picture, i.e., a draft genome assembly. This assembly is not complete by any means. In fact, most microbial genomes sequenced with only Illumina reads
will not give you a complete genome after you run an assembly using SPAdes.</p>
<p>So you are still missing a lot of pieces and thus the assemblies will yield incomplete but partially assembled DNA fragments that are known as “<strong>contigs</strong>”. They range in size from only a few hundred to hundreds of thousands of bases. You wouldn’t normally inspect them by using Word processor to count these bases. There are tools specifically designed for you to inspect these metrics, thankfully. One of these tools is known as “<strong>Quast</strong>”” and it happens to be the tool written by the same
authors as SPAdes. You will install Quast using <code class="docutils literal notranslate"><span class="pre">conda</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda search quast
conda install quast
</pre></div>
</div>
<p>Before running Quast on your assemblies, you need to first inspect what SPAdes produced after the assembly. Let’s check it out.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
<span class="nb">cd</span> /Volumes/BackupWork/teaching/BISC4234_6234/spades_assembly
ls
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
K21
K33
K55
K77
assembly_graph.fastg
assembly_graph_with_scaffolds.gfa
before_rr.fasta
contigs.fasta
contigs.paths
corrected
dataset.info
input_dataset.yaml
misc
params.txt
scaffolds.fasta
scaffolds.paths
spades.log
tmp
</pre></div></div>
</div>
<p>Now, that’s a lot of files and folders. Where do you begin? What is the most important thing for you to inspect? The only files you would need after the assembly are <code class="docutils literal notranslate"><span class="pre">contigs.fasta</span></code> and <code class="docutils literal notranslate"><span class="pre">scaffolds.fasta</span></code> files. <code class="docutils literal notranslate"><span class="pre">contigs.fasta</span></code> contains contigs (incomplete DNA fragments of the genome) of <em>Salmonella enterica</em> you just assembled. What about <code class="docutils literal notranslate"><span class="pre">scaffolds.fasta</span></code> file? It is basically similar to <code class="docutils literal notranslate"><span class="pre">contigs.fasta</span></code> file but it has scaffolding information. For example, if the SPAdes assembler has
determined that two or more contigs should be physically linked due to pairing information from original reads, then it will put them together but with “N” characters to indicate that there are missing pieces that should be filled.</p>
<p>I normally work with contig files instead of scaffold files for a number of reasons. We will get to that when we start working on genome annotation. For now, let’s use Quast to inspect the assembly metrics for the contigs. You need to first be in the folder containing the SPAdes assembly. Then type:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>quast.py contigs.fasta -o quast_contigs
</pre></div>
</div>
<p>This will run Quast on the contigs file and it will produce a folder <code class="docutils literal notranslate"><span class="pre">quast_contigs</span></code> which contains the assembly metrics. Now, let’s see what Quast found.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
<span class="nb">cd</span> /Volumes/BackupWork/teaching/BISC4234_6234/spades_assembly
cat quast_contigs/report.txt
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All statistics are based on contigs of size &gt;= 500 bp, unless otherwise noted (e.g., &#34;# contigs (&gt;= 0 bp)&#34; and &#34;Total length (&gt;= 0 bp)&#34; include all contigs).

Assembly                    contigs
# contigs (&gt;= 0 bp)         135
# contigs (&gt;= 1000 bp)      67
# contigs (&gt;= 5000 bp)      53
# contigs (&gt;= 10000 bp)     51
# contigs (&gt;= 25000 bp)     43
# contigs (&gt;= 50000 bp)     32
Total length (&gt;= 0 bp)      5097086
Total length (&gt;= 1000 bp)   5075927
Total length (&gt;= 5000 bp)   5039163
Total length (&gt;= 10000 bp)  5024899
Total length (&gt;= 25000 bp)  4900465
Total length (&gt;= 50000 bp)  4481295
# contigs                   77
Largest contig              282699
Total length                5082069
GC (%)                      52.08
N50                         170524
N75                         94555
L50                         12
L75                         24
# N&#39;s per 100 kbp           0.00
</pre></div></div>
</div>
<p>It found 135 contigs in the assembly and the total size of the assembly is about 5 Mbp, which is roughly the genome size of <em>Salmonella enterica</em>. The largest contig is more than 280 Kbp long! And G+C% of the geome is about 52%. N50 is another metric that is quite important to assess how well the assembly went. That is the number at which 50% of the assembly size is accounted for by contigs that are longer than 170,524 bases. In this example, 12 contigs can account for 50% of the genome, which
is great because fewer contigs mean it’s closer to finishing the puzzle to obtain the complete genome. Quast also produces some plots of these metrics in PDF format and an html page containing visualization of the contigs. You can open <code class="docutils literal notranslate"><span class="pre">icarus.html</span></code> file using your web brower to inspect these plots, which are interactive.</p>
</div>
<div class="section" id="Test-trimming-parameters-and-run-another-SPAdes-assembly">
<h2>Test trimming parameters and run another SPAdes assembly<a class="headerlink" href="#Test-trimming-parameters-and-run-another-SPAdes-assembly" title="Permalink to this headline">¶</a></h2>
<p>A lot of microbial genome assemblies and the qualities are impacted by how well (either aggressively or less aggressively) you trim the original raw sequence reads. Now, go back to the step where you used <code class="docutils literal notranslate"><span class="pre">bbduk.sh</span></code> command to trim these reads. Change one or two parameters to see if you can trim more aggressively (eg: increasing <code class="docutils literal notranslate"><span class="pre">qtrim</span></code> to a higher number, for example) and see if your SPAdes assembly improves or got worse. Make sure to name the assembly differently so that you are not
overwriting the previous assembly or the SPAdes assembler won’t proceed due to previous assembly folder being present. The parameter you need to change is <code class="docutils literal notranslate"><span class="pre">-o</span></code>.</p>
</div>
<div class="section" id="Cluster-login-info">
<h2>Cluster login info<a class="headerlink" href="#Cluster-login-info" title="Permalink to this headline">¶</a></h2>
<p>For next week, we will be connecting to the HPC cluster known as <code class="docutils literal notranslate"><span class="pre">cerberus</span></code> and I have put some useful information to prepare you for next week’s exercises.</p>
<p>An example command to login to this remote computer is shown below.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ssh jsaw@cerberus.colonialone.gwu.edu
</pre></div>
</div>
<p>However, before you can connect to this computer, I will need to provide the system administrator with your email information so that he can give you permission to connect to this computer. Therefore, you will need to let me know the email address you are using for your GW user id (the same as the login name used for your email account).</p>
<p>This remote computer has the following specifications:</p>
<ul class="simple">
<li><p>28 CPU cores</p></li>
<li><p>128 GB RAM</p></li>
</ul>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
        <p class="logo"><a href="index.html" title="index">
          <img class="logo" src="_static/gm_pan.png" alt="Logo"/>
        </a></p><div class="sphinx-toc sphinxlocaltoc">
    <h3><a href="index.html">Page contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Genome Assembly using SPAdes</a><ul>
<li><a class="reference internal" href="#Introduction-to-SPAdes">Introduction to SPAdes</a></li>
<li><a class="reference internal" href="#Installing-SPAdes">Installing SPAdes</a></li>
<li><a class="reference internal" href="#Running-SPAdes">Running SPAdes</a></li>
<li><a class="reference internal" href="#Checking-sequence-qualities">Checking sequence qualities</a></li>
<li><a class="reference internal" href="#Trimming-the-sequences">Trimming the sequences</a></li>
<li><a class="reference internal" href="#Run-fastqc-on-quality-trimmed-sequences-to-compare-its-quality-with-untrimmed-sequences">Run fastqc on quality-trimmed sequences to compare its quality with untrimmed sequences</a></li>
<li><a class="reference internal" href="#Running-the-assembly-on-your-laptop-or-home-computer">Running the assembly on your laptop or home computer</a></li>
<li><a class="reference internal" href="#Running-the-assembly-on-a-cluster-at-GW">Running the assembly on a cluster at GW</a></li>
<li><a class="reference internal" href="#Checking-the-quality-of-assemblies">Checking the quality of assemblies</a></li>
<li><a class="reference internal" href="#Test-trimming-parameters-and-run-another-SPAdes-assembly">Test trimming parameters and run another SPAdes assembly</a></li>
<li><a class="reference internal" href="#Cluster-login-info">Cluster login info</a></li>
</ul>
</li>
</ul>

  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/GenomeAssemblies.ipynb.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
    
    
        <div class="sidebar-toggle-group no-js">
            
            <button class="sidebar-toggle" id="sidebar-hide" title="Hide the sidebar menu">
                 «
                <span class="show-for-small">hide menu</span>
                
            </button>
            <button class="sidebar-toggle" id="sidebar-show" title="Show the sidebar menu">
                
                <span class="show-for-small">menu</span>
                <span class="hide-for-small">sidebar</span>
                 »
            </button>
        </div>
    
      <div class="clearer"></div>
    </div>
    <div class="relbar-bottom">
        
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
    <li><a href="index.html">Microbial Genomics Lab Spring 2022  documentation</a> &#187;</li>

        <li class="nav-item nav-item-this"><a href="">Genome Assembly using SPAdes</a></li> 
      </ul>
    </div>
    </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2022, Jimmy Saw.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.3.2.
    </div>
    <!-- cloud_sptheme 1.4 -->
  </body>
</html>