


<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Install tools for today’s lab &#8212; Microbial Genomics Lab Spring 2022  documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/cloud.css" />
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Noticia+Text:400,i,b,bi|Open+Sans:400,i,b,bi|Roboto+Mono:400,i,b,bi&amp;display=swap" type="text/css" />
    
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>

    
    
     
        <script src="_static/jquery.cookie.js"></script>
    

    
     
        <script src="_static/cloud.base.js"></script>
    

    
     
        <script src="_static/cloud.js"></script>
    

    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Exercises for today" href="exercises_week2.html" />
    <link rel="prev" title="Brief introduction to Markdown" href="markdown.html" /> 
        <meta name="viewport" content="width=device-width, initial-scale=1">
  </head><body>
    <div class="relbar-top">
        
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="exercises_week2.html" title="Exercises for today"
             accesskey="N">next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="markdown.html" title="Brief introduction to Markdown"
             accesskey="P">previous</a> &nbsp; &nbsp;</li>
    <li><a href="index.html">Microbial Genomics Lab Spring 2022  documentation</a> &#187;</li>

        <li class="nav-item nav-item-this"><a href="">Install tools for today’s lab</a></li> 
      </ul>
    </div>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for cloud theme */

/* nicer titles and more space for info and warning logos */

div.admonition p.admonition-title {
    background: rgba(0, 0, 0, .05);
    margin: .5em -1em;
    margin-top: -.5em !important;
    padding: .5em .5em .5em 2.65em;
}

/* indent single paragraph */
div.admonition {
    text-indent: 20px;
}
/* don't indent multiple paragraphs */
div.admonition > p {
    text-indent: 0;
}
/* remove excessive padding */
div.admonition.inline-title p.admonition-title {
    padding-left: .2em;
}
</style>
<div class="section" id="Install-tools-for-today’s-lab">
<h1>Install tools for today’s lab<a class="headerlink" href="#Install-tools-for-today’s-lab" title="Permalink to this headline">¶</a></h1>
<p>Last week, I mentioned that you should install some bioinformatic tools we will be using this week. If you did not have a chance to install them last week, please do it now. You should be able to install them all using <code class="docutils literal notranslate"><span class="pre">conda</span></code>.</p>
<div class="section" id="sra-tools">
<h2>sra-tools<a class="headerlink" href="#sra-tools" title="Permalink to this headline">¶</a></h2>
<p>Here, we will use sra-tools to download sequence data from publicly accessible databases such as NCBI. This tool allows you to download some sequence files from NCBI directly from your terminal without having to use your web browser. First, I will check if <code class="docutils literal notranslate"><span class="pre">sra-tools</span></code> is on <code class="docutils literal notranslate"><span class="pre">conda</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
conda search sra-tools
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Loading channels: ...working... done
# Name                       Version           Build  Channel
sra-tools                      2.6.2               0  bioconda
sra-tools                      2.6.3               0  bioconda
sra-tools                      2.7.0               0  bioconda
sra-tools                      2.8.0               0  bioconda
sra-tools                      2.8.1               0  bioconda
sra-tools                      2.8.2               0  bioconda
sra-tools                      2.8.2      h550f44e_1  bioconda
sra-tools                      2.9.0      h470a237_1  bioconda
sra-tools                      2.9.1      h470a237_0  bioconda
sra-tools                    2.9.1_1      h470a237_0  bioconda
sra-tools                      2.9.6      h0a44026_0  bioconda
sra-tools                     2.10.0 pl526h6de7cb9_0  bioconda
sra-tools                     2.10.1 pl526h9f37e31_0  bioconda
</pre></div></div>
</div>
<p>Now, you can install <code class="docutils literal notranslate"><span class="pre">sra-tools</span></code> by typing</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda install sra-tools
</pre></div>
</div>
<p>This will install a suite of tools you will need to download publicly available sequences from NCBI.</p>
</div>
<div class="section" id="FastQC">
<h2>FastQC<a class="headerlink" href="#FastQC" title="Permalink to this headline">¶</a></h2>
<p>FastQC is a tool you will need to use to assess sequence quality and to see if they are suitable to be used in your analyses. Earlier days, sequences produced by these so-called “Next-gen” sequencing instruments (especially in early 2010s) produced sequences with high error rates and low qualities. And they sometimes include the adapter sequences that should be removed before they can be used. To check for sequence quality, there are several tools available but one of the most well-known (and
most widely used) is FastQC. It exists in both graphical and command line versions. You can install it using <code class="docutils literal notranslate"><span class="pre">conda</span></code>. First, search for it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
conda search fastqc
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Loading channels: ...working... done
# Name                       Version           Build  Channel
fastqc                        0.10.1               0  bioconda
fastqc                        0.10.1               1  bioconda
fastqc                        0.11.2               1  bioconda
fastqc                        0.11.2      pl5.22.0_0  bioconda
fastqc                        0.11.3               0  bioconda
fastqc                        0.11.3               1  bioconda
fastqc                        0.11.4               1  bioconda
fastqc                        0.11.4               2  bioconda
fastqc                        0.11.5               1  bioconda
fastqc                        0.11.5               4  bioconda
fastqc                        0.11.5      pl5.22.0_2  bioconda
fastqc                        0.11.5      pl5.22.0_3  bioconda
fastqc                        0.11.6               2  bioconda
fastqc                        0.11.6      pl5.22.0_0  bioconda
fastqc                        0.11.6      pl5.22.0_1  bioconda
fastqc                        0.11.7               4  bioconda
fastqc                        0.11.7               5  bioconda
fastqc                        0.11.7               6  bioconda
fastqc                        0.11.7      pl5.22.0_0  bioconda
fastqc                        0.11.7      pl5.22.0_2  bioconda
fastqc                        0.11.8               0  bioconda
fastqc                        0.11.8               1  bioconda
fastqc                        0.11.8               2  bioconda
fastqc                        0.11.9               0  bioconda
</pre></div></div>
</div>
<p>As you can see, there are several versions and if you just type <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">fastqc</span></code>, it will install the latest version of it.</p>
</div>
<div class="section" id="BBmap">
<h2>BBmap<a class="headerlink" href="#BBmap" title="Permalink to this headline">¶</a></h2>
<p>FastQC will help you see what needs to be done to the original raw sequnce files but it doesn’t actually do anything else. To actually trim reads of poor quality or to discard sequences that do not meet specific requirement, you need to use tools like BBmap. It is just one of the many tools that does similar function.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
conda search bbmap
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Loading channels: ...working... done
# Name                       Version           Build  Channel
bbmap                          35.85               1  bioconda
bbmap                          35.85               2  bioconda
bbmap                          37.02               0  bioconda
bbmap                          37.10               0  bioconda
bbmap                          37.10               1  bioconda
bbmap                          37.17               0  bioconda
bbmap                          37.17               1  bioconda
bbmap                          37.52               0  bioconda
bbmap                          37.52               1  bioconda
bbmap                          37.62               0  bioconda
bbmap                          37.62               1  bioconda
bbmap                          37.66               0  bioconda
bbmap                          37.75               0  bioconda
bbmap                          37.77               0  bioconda
bbmap                          37.78               0  bioconda
bbmap                          37.90               0  bioconda
bbmap                          37.95               0  bioconda
bbmap                          37.96               0  bioconda
bbmap                          37.99               0  bioconda
bbmap                          37.99               1  bioconda
bbmap                          38.06               0  bioconda
bbmap                          38.06               2  bioconda
bbmap                          38.16               0  bioconda
bbmap                          38.18               0  bioconda
bbmap                          38.19      h470a237_0  bioconda
bbmap                          38.20      h470a237_0  bioconda
bbmap                          38.22      h1de35cc_1  bioconda
bbmap                          38.22      h470a237_0  bioconda
bbmap                          38.44      h1de35cc_0  bioconda
bbmap                          38.45      h1de35cc_0  bioconda
bbmap                          38.46      h1de35cc_0  bioconda
bbmap                          38.49      h1de35cc_0  bioconda
bbmap                          38.51      h1de35cc_0  bioconda
bbmap                          38.56      h1de35cc_0  bioconda
bbmap                          38.57      h1de35cc_0  bioconda
bbmap                          38.58      h01d97ff_0  bioconda
bbmap                         38.61b      h01d97ff_0  bioconda
bbmap                          38.62      h01d97ff_0  bioconda
bbmap                          38.63      h01d97ff_0  bioconda
bbmap                          38.65      h01d97ff_0  bioconda
bbmap                          38.67      h01d97ff_0  bioconda
bbmap                          38.68      h01d97ff_0  bioconda
bbmap                          38.69      h01d97ff_0  bioconda
bbmap                          38.70      h01d97ff_0  bioconda
bbmap                          38.71      h01d97ff_0  bioconda
bbmap                          38.72      h01d97ff_0  bioconda
bbmap                          38.73      h01d97ff_0  bioconda
bbmap                          38.75      h01d97ff_0  bioconda
bbmap                          38.76      h01d97ff_0  bioconda
bbmap                          38.79      h01d97ff_0  bioconda
bbmap                          38.84      h0b31af3_0  bioconda
bbmap                          38.84      hf29c6f4_1  bioconda
bbmap                          38.86      hf29c6f4_0  bioconda
</pre></div></div>
</div>
<p>This is a collection of tools that you will need to use to remove poor-quality sequence reads and to trim bad or contaminated sequences from the pool. Install it by typing:</p>
<p><code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">bbmap</span></code></p>
<p>Now you have all the tools we will need to use for today’s exercise.</p>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
        <p class="logo"><a href="index.html" title="index">
          <img class="logo" src="_static/gm_pan.png" alt="Logo"/>
        </a></p><div class="sphinx-toc sphinxlocaltoc">
    <h3><a href="index.html">Page contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Install tools for today’s lab</a><ul>
<li><a class="reference internal" href="#sra-tools">sra-tools</a></li>
<li><a class="reference internal" href="#FastQC">FastQC</a></li>
<li><a class="reference internal" href="#BBmap">BBmap</a></li>
</ul>
</li>
</ul>

  </div>
  <div class="sphinxprev">
    <h4>Previous page</h4>
    <p class="topless"><a href="markdown.html"
                          title="Previous page">&larr; Brief introduction to Markdown</a></p>
  </div>
  <div class="sphinxnext">
    <h4>Next page</h4>
    <p class="topless"><a href="exercises_week2.html"
                          title="Next page">&rarr; Exercises for today</a></p>
  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/tools_week2.ipynb.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
    
    
        <div class="sidebar-toggle-group no-js">
            
            <button class="sidebar-toggle" id="sidebar-hide" title="Hide the sidebar menu">
                 «
                <span class="show-for-small">hide menu</span>
                
            </button>
            <button class="sidebar-toggle" id="sidebar-show" title="Show the sidebar menu">
                
                <span class="show-for-small">menu</span>
                <span class="hide-for-small">sidebar</span>
                 »
            </button>
        </div>
    
      <div class="clearer"></div>
    </div>
    <div class="relbar-bottom">
        
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="exercises_week2.html" title="Exercises for today"
             >next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="markdown.html" title="Brief introduction to Markdown"
             >previous</a> &nbsp; &nbsp;</li>
    <li><a href="index.html">Microbial Genomics Lab Spring 2022  documentation</a> &#187;</li>

        <li class="nav-item nav-item-this"><a href="">Install tools for today’s lab</a></li> 
      </ul>
    </div>
    </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2022, Jimmy Saw.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.3.2.
    </div>
    <!-- cloud_sptheme 1.4 -->
  </body>
</html>